### PR TITLE
SSOC: KeyAuthAI: Keystroke Dynamics-Based User Authentication

### LINKED ISSUE #438: SSOC: KeyAuthAI: Keystroke Dynamics-Based User Authentication
<h2>🕵️ KeyAuthAI: Keystroke Dynamics-Based User Authentication</h2>
<h3>📌 Project Overview</h3>
<p><strong>KeyAuthAI</strong> is a behavioral biometrics system that authenticates users based on their <strong>typing patterns</strong>—including speed, dwell time, flight time, and key sequences. The model builds a <strong>unique typing profile</strong> per user and can <strong>detect imposters or anomalies</strong> using supervised or unsupervised learning.</p>
<p>✅ All data is collected live from user input — no external dataset needed.</p>
<hr>
<h3>💡 Use Cases</h3>
<ul>
<li>
<p>🛡️ Invisible 2FA for sensitive applications</p>
</li>
<li>
<p>🧠 Cognitive profiling or user behavior tracking</p>
</li>
<li>
<p>🖥️ Continuous authentication in enterprise systems</p>
</li>
</ul>
<hr>
<h3>📂 Folder Structure</h3>
<pre><code>keyauthai/
│
├── data/
│   └── keystroke_logger.py       # Captures timing, sequence, etc.
│   └── user_data.json            # Stores user profiles &amp; raw metrics
│
├── features/
│   └── extractor.py              # Calculates dwell, flight, and n-gram timings
│
├── model/
│   ├── train_model.py            # Train classifier or anomaly detector
│   ├── model_svm.pkl             # Saved model (SVM, KNN, etc.)
│   └── verify_user.py            # Authentication logic
│
├── ui/
│   └── auth_terminal.py          # CLI/GUI login prompt with keystroke capture
│
├── utils/
│   └── metrics.py                # FAR, FRR, EER, ROC plotting
│
├── tests/
│   └── test_data_collection.py
│
├── requirements.txt
├── README.md
└── LICENSE
</code></pre>
<hr>
<h3>🧠 What It Measures</h3>

Metric | Description
-- | --
Dwell Time | Time a key is held down
Flight Time | Time between releasing one key and pressing another
Typing Speed | Chars per second
N-gram Patterns | Common sequences like "th", "ing", etc.

<hr>
<h3>🔐 Authentication Flow</h3>
<ol>
<li>
<p><strong>Registration Phase</strong>:</p>
<ul>
<li>
<p>User types a known passphrase 5–10 times.</p>
</li>
<li>
<p>Model learns statistical patterns of keystroke dynamics.</p>
</li>
</ul>
</li>
<li>
<p><strong>Verification Phase</strong>:</p>
<ul>
<li>
<p>Same passphrase typed once → model checks if it's consistent with stored profile.</p>
</li>
</ul>
</li>
</ol>
<hr>
<h3>🧠 Model Options</h3>
<ul>
<li>
<p>🟢 Supervised: SVM, Random Forest, KNN</p>
</li>
<li>
<p>🔵 Unsupervised: One-Class SVM, Isolation Forest (good for 1-user systems)</p>
</li>
<li>
<p>🧪 Optional: LSTM for temporal sequence analysis (advanced)</p>
</li>
</ul>
<hr>
<h3>🧪 Metrics</h3>
<ul>
<li>
<p>FAR (False Acceptance Rate)</p>
</li>
<li>
<p>FRR (False Rejection Rate)</p>
</li>
<li>
<p>EER (Equal Error Rate — ideal for tuning)</p>
</li>
</ul>
<hr>
<h3>🔧 Features</h3>
<ul>
<li>
<p>👨‍💻 Local biometric authentication system</p>
</li>
<li>
<p>🎯 Lightweight, real-time typing data collection</p>
</li>
<li>
<p>📈 Track improvement in accuracy over multiple sessions</p>
</li>
<li>
<p>📦 Pluggable with login forms or CLI tools</p>
</li>
<li>
<p>🔐 No passwords stored — only behavior!</p>
</li>
</ul>
<hr>
<h3>📘 README Highlights</h3>
<p><strong>How it works:</strong></p>
<ul>
<li>
<p>Captures keystroke timing while the user types.</p>
</li>
<li>
<p>Extracts behavioral features.</p>
</li>
<li>
<p>Trains a model to match future typing patterns with the known profile.</p>
</li>
</ul>
<p><strong>Technologies:</strong></p>
<ul>
<li>
<p>Python</p>
</li>
<li>
<p><code inline="">pynput</code> or <code inline="">keyboard</code> for keylogging</p>
</li>
<li>
<p>Scikit-learn or PyTorch</p>
</li>
<li>
<p>Matplotlib for visual metrics</p>
</li>
</ul>
<hr>
<h3>🧪 Optional Enhancements</h3>
<ul>
<li>
<p>Use mobile touchscreen pressure data if available (via JS or Android sensors)</p>
</li>
<li>
<p>Add continuous monitoring (not just login)</p>
</li>
<li>
<p>Compare with typing of known imposters or bots</p>
</li>
<li>
<p>Export model to ONNX for web use</p>

### PR DESCRIPTION
**Pull request by SK8-infi**
Closes #438 

<h2>🕵️ KeyAuthAI: Keystroke Dynamics-Based User Authentication</h2>
<h3>📌 Project Overview</h3>
<p><strong>KeyAuthAI</strong> is a behavioral biometrics system that authenticates users based on their <strong>typing patterns</strong>—including speed, dwell time, flight time, and key sequences. The model builds a <strong>unique typing profile</strong> per user and can <strong>detect imposters or anomalies</strong> using supervised or unsupervised learning.</p>
<p>✅ All data is collected live from user input — no external dataset needed.</p>
<hr>
<h3>💡 Use Cases</h3>
<ul>
<li>
<p>🛡️ Invisible 2FA for sensitive applications</p>
</li>
<li>
<p>🧠 Cognitive profiling or user behavior tracking</p>
</li>
<li>
<p>🖥️ Continuous authentication in enterprise systems</p>
</li>
</ul>
<hr>
<h3>📂 Folder Structure</h3>
<pre><code>keyauthai/
│
├── data/
│   └── keystroke_logger.py       # Captures timing, sequence, etc.
│   └── user_data.json            # Stores user profiles &amp; raw metrics
│
├── features/
│   └── extractor.py              # Calculates dwell, flight, and n-gram timings
│
├── model/
│   ├── train_model.py            # Train classifier or anomaly detector
│   ├── model_svm.pkl             # Saved model (SVM, KNN, etc.)
│   └── verify_user.py            # Authentication logic
│
├── ui/
│   └── auth_terminal.py          # CLI/GUI login prompt with keystroke capture
│
├── utils/
│   └── metrics.py                # FAR, FRR, EER, ROC plotting
│
├── tests/
│   └── test_data_collection.py
│
├── requirements.txt
├── README.md
└── LICENSE
</code></pre>
<hr>
<h3>🧠 What It Measures</h3>

Metric | Description
-- | --
Dwell Time | Time a key is held down
Flight Time | Time between releasing one key and pressing another
Typing Speed | Chars per second
N-gram Patterns | Common sequences like "th", "ing", etc.

<hr>
<h3>🔐 Authentication Flow</h3>
<ol>
<li>
<p><strong>Registration Phase</strong>:</p>
<ul>
<li>
<p>User types a known passphrase 5–10 times.</p>
</li>
<li>
<p>Model learns statistical patterns of keystroke dynamics.</p>
</li>
</ul>
</li>
<li>
<p><strong>Verification Phase</strong>:</p>
<ul>
<li>
<p>Same passphrase typed once → model checks if it's consistent with stored profile.</p>
</li>
</ul>
</li>
</ol>
<hr>
<h3>🧠 Model Options</h3>
<ul>
<li>
<p>🟢 Supervised: SVM, Random Forest, KNN</p>
</li>
<li>
<p>🔵 Unsupervised: One-Class SVM, Isolation Forest (good for 1-user systems)</p>
</li>
<li>
<p>🧪 Optional: LSTM for temporal sequence analysis (advanced)</p>
</li>
</ul>
<hr>
<h3>🧪 Metrics</h3>
<ul>
<li>
<p>FAR (False Acceptance Rate)</p>
</li>
<li>
<p>FRR (False Rejection Rate)</p>
</li>
<li>
<p>EER (Equal Error Rate — ideal for tuning)</p>
</li>
</ul>
<hr>
<h3>🔧 Features</h3>
<ul>
<li>
<p>👨‍💻 Local biometric authentication system</p>
</li>
<li>
<p>🎯 Lightweight, real-time typing data collection</p>
</li>
<li>
<p>📈 Track improvement in accuracy over multiple sessions</p>
</li>
<li>
<p>📦 Pluggable with login forms or CLI tools</p>
</li>
<li>
<p>🔐 No passwords stored — only behavior!</p>
</li>
</ul>
<hr>
<h3>📘 README Highlights</h3>
<p><strong>How it works:</strong></p>
<ul>
<li>
<p>Captures keystroke timing while the user types.</p>
</li>
<li>
<p>Extracts behavioral features.</p>
</li>
<li>
<p>Trains a model to match future typing patterns with the known profile.</p>
</li>
</ul>
<p><strong>Technologies:</strong></p>
<ul>
<li>
<p>Python</p>
</li>
<li>
<p><code inline="">pynput</code> or <code inline="">keyboard</code> for keylogging</p>
</li>
<li>
<p>Scikit-learn or PyTorch</p>
</li>
<li>
<p>Matplotlib for visual metrics</p>
</li>
</ul>
<hr>
<h3>🧪 Optional Enhancements</h3>
<ul>
<li>
<p>Use mobile touchscreen pressure data if available (via JS or Android sensors)</p>
</li>
<li>
<p>Add continuous monitoring (not just login)</p>
</li>
<li>
<p>Compare with typing of known imposters or bots</p>
</li>
<li>
<p>Export model to ONNX for web use</p>

### CODE REVIEW

**Comment by SakuraPa78** [comment]
Great project 
i am interest
